framework/src/wifi_manager/wifi_manager.c

```diff
@@ -393,119 +396,184 @@ wifi_manager_result_e wifi_manager_init(wifi_manager_cb_s *wmcb)
+			ndbg("w_info_mutex_create fail");
+			goto error_without_mutex_release;
+		}
+	}
+
+	wifi_mutex_acquire(w_info_mutex, WIFI_UTILS_FOREVER);
 	if (w_mutex != NULL) {
 		ndbg("WI-FI is already initialized.\n");
+		wifi_mutex_release(w_info_mutex);
 		return WIFI_MANAGER_INITIALIZED;
 	}
 
 	w_mutex = (wifi_mutex *)malloc(sizeof(wifi_mutex));
 	if (w_mutex == NULL) {
 		ndbg("w_mutex malloc fails.\n");
-		return WIFI_MANAGER_FAIL;
+		wifi_mutex_release(w_info_mutex);
+		goto error_with_mutex_destory;
 	}
 
 	result = wifi_mutex_create(w_mutex);
 	if (result != WIFI_UTILS_SUCCESS) {
 		ndbg("wifi_mutex_create fail");
-		goto error_without_mutex_release;
+		wifi_mutex_release(w_info_mutex);
+		goto error_with_mutex_destory;
 	}
 
-	wifi_mutex_acquire(w_mutex, WIFI_UTILS_FOREVER);
+	/* wifi_manager mode is switched to wifi_initializing */
+	g_manager_info.mode = WIFI_INITIALIZING;
+	wifi_mutex_release(w_info_mutex);
 
-	result = wifi_utils_init();
+#ifdef CONFIG_ENABLE_IOTIVITY
+	struct mq_attr lq_attr;
+	lq_attr.mq_maxmsg = 10;
+	lq_attr.mq_msgsize = 4;
+	lq_attr.mq_flags = 0;
+	g_dw_nwevent_mqfd = mq_open("netlink_evtq", O_WRONLY | O_CREAT, 0666, &lq_attr);
 
+	if (g_dw_nwevent_mqfd == (mqd_t)ERROR) {
+		ndbg("iotivity connect event message queue init fail");
+		goto error_with_mutex_destory_all;
+	}
+#endif
+
+	wifi_mutex_acquire(w_mutex, WIFI_UTILS_FOREVER);
+	result = wifi_utils_init();
 	if (result != WIFI_UTILS_SUCCESS) {
 		ndbg("wifi_utils_init fail");
 		goto error_with_mutex_release;
 	}
 
 	wifi_utils_register_callback(wifi_linkup_event_func, wifi_linkdown_event_func);
+
+	if (wifi_utils_get_info(&info) != WIFI_UTILS_SUCCESS) {
+		ndbg("wifi_utils_get_info fail");
+		goto error_with_mutex_release;
+	}
+	wifi_mutex_release(w_mutex);
+
+	/* update wifi_manager info by being protected by w_info_mutex */
+	wifi_mutex_acquire(w_info_mutex, WIFI_UTILS_FOREVER);
 	g_manager_info.ssid[0] = '\0';
 	g_manager_info.ip4_address[0] = '\0';
 	g_manager_info.rssi = 0;
 	g_manager_info.mode = STA_MODE;
 	g_manager_callback = wmcb;
 	wifi_status_set(AP_DISCONNECTED);
-
-	if (wifi_utils_get_info(&info) != WIFI_UTILS_SUCCESS) {
-		ndbg("wifi_utils_get_info fail");
-		goto error_with_mutex_release;
-	}
 	strncpy(g_manager_info.mac_address, info.mac_address, 6);
+	wifi_mutex_release(w_info_mutex);
+		
+	return WIFI_MANAGER_SUCCESS;
 
+error_with_mutex_release:
 	wifi_mutex_release(w_mutex);
 
 #ifdef CONFIG_ENABLE_IOTIVITY
-	struct mq_attr lq_attr;
-	lq_attr.mq_maxmsg = 10;
-	lq_attr.mq_msgsize = 4;
-	lq_attr.mq_flags = 0;
-	g_dw_nwevent_mqfd = mq_open("netlink_evtq", O_WRONLY | O_CREAT, 0666, &lq_attr);
-
-	if (g_dw_nwevent_mqfd == (mqd_t)ERROR) {
-		ndbg("iotivity connect event message queue init fail");
-		goto error_with_mutex_destory;
-	}
+error_with_mutex_destory_all:
 #endif
+	wifi_mutex_acquire(w_info_mutex, WIFI_UTILS_FOREVER);
+	g_manager_info.mode = WIFI_NONE;
 
-	return WIFI_MANAGER_SUCCESS;
+	result = wifi_mutex_destroy(w_mutex);
```

Comment:
does w_mutex need to be destroyed inside w_info_mutex

Reply:
w_mutex is created when g_manager_info.mode is switched to 'STA_MODE', that is, initialization is successful and is destroyed when g_manager_info.mode is switched to 'WIFI_NONE', that is, deinitialization is successful.
This is why w_info_mutex protects w_mutex creation and deletion.